Using a GameSession table in the db to implement the CRU functionalities. Game in working condition.

Emmett-42 opened the pull request:
resolves Issue #1

Emmett-42 committed:
deleted previous games table

Emmett-42 committed:
Merge branch 'main' of https://github.com/Ayesha-42/HangmanRails into db-table

Emmett-42 committed:
creating a record with the word_to_guess in the table GameSessions

Emmett-42 committed:
reads the current game session record from the table and updates the guesses string with the letter inputted by the user. Accordingly, returns the hidden_word with the letter revealed if exists.

Emmett-42 committed:
calculates lives remaining from checking all the wrong letters

From the current game session record retrived from the table, it iterates through the guesses string of letters to check which are incorrect, i.e. not present is the word_to_guess. According to this it keeps deducting from the total lives to return the remaining lives to the controller to display in the view. The controller also checks the losing condition, that is when the user has no more lives left to guess. In this case, it returns the result page with the appropriate message and the word revealed.
:wq

Emmett-42 committed:
win condition check added. Result page has a restart game option.

piratechicken commented on app/controllers/game_controller.rb:
```diff
@@ -1,35 +1,38 @@
-      @message = 'You won the game!'
-      render('result')
+    @lives = game.lives
+    if game.win?
```
I reckon two improvements to this. 

1. GameSession determines if it's a win or not. So let's also have it determine if we lose. In the GameSession class:
```ruby
def lose
  lives < 1
end
```
2. I think this would be better as an if/elsif statement, to emphasise that only one option of win/lose/continue is possible. Then you don't need the `return` either.
```ruby
if game.win?
  #
elsif game.lose?
  #
else
 # 
end
```

piratechicken commented on app/controllers/game_controller.rb:
```diff
@@ -1,35 +1,38 @@
-      render('result')
+    @lives = game.lives
+    if game.win?
+      redirect_to(:action => 'result', :message => "Yay! you won the game :)", :word => game.word_to_guess)
```
Far more common to use the new hash syntax these days:

```suggestion
      redirect_to(action: 'result', message: 'Yay! you won the game :)', word: game.word_to_guess)
```

EDIT: I think rubocop will ping you on this actually

Commit:
run bundle install to get nokogiri gem version for mac (darwin os)

piratechicken commented on Gemfile.lock:
```diff
@@ -132,6 +132,8 @@ GEM
     nio4r (2.5.8)
     nokogiri (1.13.6-x64-mingw-ucrt)
       racc (~> 1.4)
+    nokogiri (1.13.6-x86_64-darwin)
```
This is me running bundle install for mac

Emmett-42 replied:
Didn't consider this possibility. Thanks for making the inclusion.

piratechicken commented on app/models/game_session.rb:
```diff
@@ -0,0 +1,30 @@
+class GameSession < ApplicationRecord
+  LIVES = 5
+
+  def lives
```
I'd like to keep `LIVES` and `lives` as difference names, just to make it really clear to the reader about what they are.

Maybe `LIVES` could be `STARTING_LIVES`
and `lives` could be `lives_remaining`
?

piratechicken commented on app/controllers/game_controller.rb:
```diff
@@ -1,35 +1,38 @@
-    @word_to_guess ||= @words.sample
-    @hidden_word = @word_to_guess.chars.map { '_' }
+  def start
+    words = %w[programming fun internship ruby rails coding design]
+    game = GameSession.new(word_to_guess: words.sample, guesses: '')
```
DO you think this logic belongs here, or should it live in GameSession? There's a hook I can show you for this if you want.

piratechicken reviewed (changes requested):
Great stuff! It worked!!

I've left a couple of nitpicky comments and I think you need to run rubocop over this before we merge.

piratechicken commented on app/controllers/game_controller.rb:
```diff
@@ -1,35 +1,38 @@
 class GameController < ApplicationController
```
Should this be GameSessionController?

Emmett-42 replied:
Thanks for pointing this out. Have made the relevant class name changes and accordingly updated the 'routes' file for correct navigation calling.

Emmett-42 committed:
Corrections made to follow better conventions and design principles
rubocop offenses addressed
win-loss conditions shifted to model class to just be checked within a if-elsif block to render appropriate page in the controller
improving the naming and readibility of the variable and methods relating to 'lives'
using a hook within the model class to initialize the record in the table

Emmett-42 committed:
file renamed for better naming convention

piratechicken commented on app/models/game_session.rb:
```diff
@@ -0,0 +1,38 @@
+
+  before_create :set_word
+
+  def set_word
+    self.word_to_guess = WORDS.sample
+  end
```
I think we can move this to private, since it's not something we want any outside class to have access to at the moment.

piratechicken reviewed (commented):
One final comment!

Emmett-42 committed:
proper access specifier implemented to keep the relevant method private in the model class.

piratechicken reviewed (approved):
Awesome, thanks!
